File: src/cpp/abx2xml/abx2xml.cc
#include "abx2xml.hpp"
#include <fstream>
#include <sstream>
#include <iomanip>
#include <cstring>
#include <cmath>
#include <array>
#include <algorithm>
#include <bit>

namespace abx2xml {


std::string encode_xml_entities(std::string_view text) {
    std::string result;
    result.reserve(text.size());
    
    for (char c : text) {
        switch (c) {
            case '&': result += "&amp;"; break;
            case '<': result += "&lt;"; break;
            case '>': result += "&gt;"; break;
            case '"': result += "&quot;"; break;
            case '\'': result += "&apos;"; break;
            default: result += c; break;
        }
    }
    
    return result;
}


std::string hex_encode(const uint8_t* data, size_t len) {
    std::ostringstream oss;
    oss << std::hex << std::setfill('0');
    
    for (size_t i = 0; i < len; ++i) {
        oss << std::setw(2) << static_cast<int>(data[i]);
    }
    
    return oss.str();
}


std::string base64_encode(const uint8_t* data, size_t len) {
    static constexpr std::string_view b64_table = 
        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/";
    
    std::string result;
    result.reserve(((len + 2) / 3) * 4);
    
    for (size_t i = 0; i < len; i += 3) {
        uint32_t n = static_cast<uint32_t>(data[i]) << 16;
        if (i + 1 < len) n |= static_cast<uint32_t>(data[i + 1]) << 8;
        if (i + 2 < len) n |= static_cast<uint32_t>(data[i + 2]);
        
        result += b64_table[(n >> 18) & 0x3F];
        result += b64_table[(n >> 12) & 0x3F];
        result += (i + 1 < len) ? b64_table[(n >> 6) & 0x3F] : '=';
        result += (i + 2 < len) ? b64_table[n & 0x3F] : '=';
    }
    
    return result;
}


DataInput::DataInput(std::istream& input) 
    : reader(input), peeked_byte(std::nullopt) {}

uint8_t DataInput::read_byte() {
    if (peeked_byte.has_value()) {
        uint8_t byte = *peeked_byte;
        peeked_byte.reset();
        return byte;
    }
    
    char c;
    if (!reader.get(c)) {
        throw AbxError("Failed to read byte");
    }
    return static_cast<uint8_t>(c);
}

uint8_t DataInput::peek_byte() {
    if (peeked_byte.has_value()) {
        return *peeked_byte;
    }
    
    peeked_byte = read_byte();
    return *peeked_byte;
}

uint16_t DataInput::read_short() {
    std::array<uint8_t, 2> bytes = {read_byte(), read_byte()};
    
    if constexpr (std::endian::native == std::endian::little) {
        std::reverse(bytes.begin(), bytes.end());
    }
    
    uint16_t result;
    std::memcpy(&result, bytes.data(), sizeof(result));
    return result;
}

int32_t DataInput::read_int() {
    std::array<uint8_t, 4> bytes = {
        read_byte(), read_byte(), read_byte(), read_byte()
    };
    
    if constexpr (std::endian::native == std::endian::little) {
        std::reverse(bytes.begin(), bytes.end());
    }
    
    int32_t result;
    std::memcpy(&result, bytes.data(), sizeof(result));
    return result;
}

int64_t DataInput::read_long() {
    std::array<uint8_t, 8> bytes = {
        read_byte(), read_byte(), read_byte(), read_byte(),
        read_byte(), read_byte(), read_byte(), read_byte()
    };
    
    if constexpr (std::endian::native == std::endian::little) {
        std::reverse(bytes.begin(), bytes.end());
    }
    
    int64_t result;
    std::memcpy(&result, bytes.data(), sizeof(result));
    return result;
}

float DataInput::read_float() {
    int32_t int_val = read_int();
    uint32_t uint_val = static_cast<uint32_t>(int_val);
    
    float result;
    std::memcpy(&result, &uint_val, sizeof(float));
    return result;
}

double DataInput::read_double() {
    int64_t long_val = read_long();
    uint64_t ulong_val = static_cast<uint64_t>(long_val);
    
    double result;
    std::memcpy(&result, &ulong_val, sizeof(double));
    return result;
}

std::string DataInput::read_utf() {
    uint16_t length = read_short();
    std::string result(length, '\0');
    
    for (uint16_t i = 0; i < length; ++i) {
        result[i] = static_cast<char>(read_byte());
    }
    
    return result;
}

std::string DataInput::read_interned_utf() {
    uint16_t index = read_short();
    
    if (index == 0xFFFF) {
        std::string str = read_utf();
        interned_strings.push_back(str);
        return str;
    }
    
    if (index >= interned_strings.size()) {
        throw AbxError("Invalid interned string index: " + std::to_string(index));
    }
    
    return interned_strings[index];
}

std::vector<uint8_t> DataInput::read_bytes(uint16_t length) {
    std::vector<uint8_t> data(length);
    
    for (uint16_t i = 0; i < length; ++i) {
        data[i] = read_byte();
    }
    
    return data;
}


BinaryXmlDeserializer::BinaryXmlDeserializer(std::istream& reader, std::ostream& writer)
    : input(reader), output(writer) {
    
    std::array<uint8_t, 4> magic;
    for (auto& byte : magic) {
        char c;
        if (!reader.get(c)) {
            throw AbxError("Failed to read magic header");
        }
        byte = static_cast<uint8_t>(c);
    }
    
    if (!std::equal(magic.begin(), magic.end(), std::begin(PROTOCOL_MAGIC_VERSION_0))) {
        throw AbxError("Invalid ABX file format - magic header mismatch");
    }
}

void BinaryXmlDeserializer::deserialize() {
    output << R"(<?xml version="1.0" encoding="UTF-8"?>)";
    
    while (true) {
        try {
            if (!process_token()) {
                break;
            }
        } catch (const AbxError&) {
            break;
        }
    }
}

bool BinaryXmlDeserializer::process_token() {
    uint8_t token = input.read_byte();
    uint8_t command = token & 0x0F;
    uint8_t type_info = token & 0xF0;
    
    switch (command) {
        case START_DOCUMENT:
            return true;
            
        case END_DOCUMENT:
            return false;
            
        case START_TAG: {
            auto tag_name = input.read_interned_utf();
            output << '<' << tag_name;
            
            while (true) {
                try {
                    uint8_t next_token = input.peek_byte();
                    if ((next_token & 0x0F) == ATTRIBUTE) {
                        
                        (void)input.read_byte();  
                        process_attribute(next_token);
                    } else {
                        break;
                    }
                } catch (const AbxError&) {
                    break;
                }
            }
            
            output << '>';
            return true;
        }
        
        case END_TAG: {
            auto tag_name = input.read_interned_utf();
            output << "</" << tag_name << '>';
            return true;
        }
        
        case TEXT:
            if (type_info == TYPE_STRING) {
                auto text = input.read_utf();
                if (!text.empty()) {
                    output << encode_xml_entities(text);
                }
            }
            return true;
            
        case CDSECT:
            if (type_info == TYPE_STRING) {
                auto text = input.read_utf();
                output << "<![CDATA[" << text << "]]>";
            }
            return true;
            
        case COMMENT:
            if (type_info == TYPE_STRING) {
                auto text = input.read_utf();
                output << "<!--" << text << "-->";
            }
            return true;
            
        case PROCESSING_INSTRUCTION:
            if (type_info == TYPE_STRING) {
                auto text = input.read_utf();
                output << "<?" << text << "?>";
            }
            return true;
            
        case DOCDECL:
            if (type_info == TYPE_STRING) {
                auto text = input.read_utf();
                output << "<!DOCTYPE " << text << '>';
            }
            return true;
            
        case ENTITY_REF:
            if (type_info == TYPE_STRING) {
                auto text = input.read_utf();
                output << '&' << text << ';';
            }
            return true;
            
        case IGNORABLE_WHITESPACE:
            if (type_info == TYPE_STRING) {
                auto text = input.read_utf();
                output << text;
            }
            return true;
            
        default:
            return true;
    }
}

void BinaryXmlDeserializer::process_attribute(uint8_t token) {
    uint8_t type_info = token & 0xF0;
    auto name = input.read_interned_utf();
    
    output << ' ' << name << "=\"";
    
    switch (type_info) {
        case TYPE_STRING: {
            auto value = input.read_utf();
            output << encode_xml_entities(value);
            break;
        }
        
        case TYPE_STRING_INTERNED: {
            auto value = input.read_interned_utf();
            output << encode_xml_entities(value);
            break;
        }
        
        case TYPE_INT: {
            int32_t value = input.read_int();
            output << value;
            break;
        }
        
        case TYPE_INT_HEX: {
            int32_t value = input.read_int();
            if (value == -1) {
                output << value;
            } else {
                output << std::hex << static_cast<uint32_t>(value) << std::dec;
            }
            break;
        }
        
        case TYPE_LONG: {
            int64_t value = input.read_long();
            output << value;
            break;
        }
        
        case TYPE_LONG_HEX: {
            int64_t value = input.read_long();
            if (value == -1) {
                output << value;
            } else {
                output << std::hex << static_cast<uint64_t>(value) << std::dec;
            }
            break;
        }
        
        case TYPE_FLOAT: {
            float value = input.read_float();
            if (std::floor(value) == value && std::isfinite(value)) {
                output << std::fixed << std::setprecision(1) << value;
            } else {
                output << value;
            }
            output << std::defaultfloat;
            break;
        }
        
        case TYPE_DOUBLE: {
            double value = input.read_double();
            if (std::floor(value) == value && std::isfinite(value)) {
                output << std::fixed << std::setprecision(1) << value;
            } else {
                output << value;
            }
            output << std::defaultfloat;
            break;
        }
        
        case TYPE_BOOLEAN_TRUE:
            output << "true";
            break;
            
        case TYPE_BOOLEAN_FALSE:
            output << "false";
            break;
            
        case TYPE_BYTES_HEX: {
            uint16_t length = input.read_short();
            auto bytes = input.read_bytes(length);
            output << hex_encode(bytes.data(), bytes.size());
            break;
        }
        
        case TYPE_BYTES_BASE64: {
            uint16_t length = input.read_short();
            auto bytes = input.read_bytes(length);
            output << base64_encode(bytes.data(), bytes.size());
            break;
        }
        
        default:
            throw AbxError("Unknown attribute type: " + std::to_string(type_info));
    }
    
    output << '"';
}


void AbxToXmlConverter::convert(std::istream& input, std::ostream& output) {
    BinaryXmlDeserializer deserializer(input, output);
    deserializer.deserialize();
}

void AbxToXmlConverter::convert_file(const std::filesystem::path& input_path, 
                                     const std::filesystem::path& output_path) {
    if (input_path == output_path) {
        
        std::ifstream infile(input_path, std::ios::binary);
        if (!infile) {
            throw AbxError("Failed to open input file: " + input_path.string());
        }
        
        std::ostringstream buffer;
        convert(infile, buffer);
        infile.close();
        
        std::ofstream outfile(output_path, std::ios::binary);
        if (!outfile) {
            throw AbxError("Failed to open output file: " + output_path.string());
        }
        outfile << buffer.str();
    } else {
        std::ifstream infile(input_path, std::ios::binary);
        if (!infile) {
            throw AbxError("Failed to open input file: " + input_path.string());
        }
        
        std::ofstream outfile(output_path, std::ios::binary);
        if (!outfile) {
            throw AbxError("Failed to open output file: " + output_path.string());
        }
        
        convert(infile, outfile);
    }
}

void AbxToXmlConverter::convert_stdin_stdout() {
    convert(std::cin, std::cout);
}

void AbxToXmlConverter::convert_stdin_to_file(const std::filesystem::path& output_path) {
    std::ofstream outfile(output_path, std::ios::binary);
    if (!outfile) {
        throw AbxError("Failed to open output file: " + output_path.string());
    }
    convert(std::cin, outfile);
}

void AbxToXmlConverter::convert_file_to_stdout(const std::filesystem::path& input_path) {
    std::ifstream infile(input_path, std::ios::binary);
    if (!infile) {
        throw AbxError("Failed to open input file: " + input_path.string());
    }
    convert(infile, std::cout);
}

std::string AbxToXmlConverter::convert_bytes(const uint8_t* data, size_t length) {
    std::string input_str(reinterpret_cast<const char*>(data), length);
    std::istringstream input(input_str);
    std::ostringstream output;
    convert(input, output);
    return output.str();
}

std::string AbxToXmlConverter::convert_string(std::string_view abx_data) {
    return convert_bytes(reinterpret_cast<const uint8_t*>(abx_data.data()), abx_data.size());
}

} 

File: src/abx2xml.rs
use android_xml_converter::*;
use base64::Engine;
use faster_hex::hex_string;
use std::env;
use std::fs::File;
use std::io::{self, BufReader, BufWriter, Cursor, Read, Write};

// ============================================================================
// Data Input Reader
// ============================================================================

pub struct DataInput<R: Read> {
    reader: R,
    interned_strings: Vec<String>,
    peeked_byte: Option<u8>,
}

impl<R: Read> DataInput<R> {
    pub fn new(reader: R) -> Self {
    Self {
        reader,
        interned_strings: Vec::with_capacity(INITIAL_STRING_POOL_CAPACITY),
        peeked_byte: None,
    }
}

    pub fn read_byte(&mut self) -> Result<u8> {
        if let Some(byte) = self.peeked_byte.take() {
            return Ok(byte);
        }
        let mut buf = [0u8; 1];
        self.reader
            .read_exact(&mut buf)
            .map_err(|_| ConversionError::ReadError("byte".to_string()))?;
        Ok(buf[0])
    }

    pub fn peek_byte(&mut self) -> Result<u8> {
        if let Some(byte) = self.peeked_byte {
            return Ok(byte);
        }
        let byte = self.read_byte()?;
        self.peeked_byte = Some(byte);
        Ok(byte)
    }

    pub fn read_short(&mut self) -> Result<u16> {
        let mut buf = [0u8; 2];
        if let Some(byte) = self.peeked_byte.take() {
            buf[0] = byte;
            self.reader
                .read_exact(&mut buf[1..])
                .map_err(|_| ConversionError::ReadError("short".to_string()))?;
        } else {
            self.reader
                .read_exact(&mut buf)
                .map_err(|_| ConversionError::ReadError("short".to_string()))?;
        }
        Ok(u16::from_be_bytes(buf))
    }

    pub fn read_int(&mut self) -> Result<i32> {
        let mut buf = [0u8; 4];
        let start_idx = if let Some(byte) = self.peeked_byte.take() {
            buf[0] = byte;
            1
        } else {
            0
        };
        self.reader
            .read_exact(&mut buf[start_idx..])
            .map_err(|_| ConversionError::ReadError("int".to_string()))?;
        Ok(i32::from_be_bytes(buf))
    }

    pub fn read_long(&mut self) -> Result<i64> {
        let mut buf = [0u8; 8];
        let start_idx = if let Some(byte) = self.peeked_byte.take() {
            buf[0] = byte;
            1
        } else {
            0
        };
        self.reader
            .read_exact(&mut buf[start_idx..])
            .map_err(|_| ConversionError::ReadError("long".to_string()))?;
        Ok(i64::from_be_bytes(buf))
    }

    pub fn read_float(&mut self) -> Result<f32> {
        let int_value = self.read_int()? as u32;
        Ok(f32::from_bits(int_value))
    }

    pub fn read_double(&mut self) -> Result<f64> {
        let int_value = self.read_long()? as u64;
        Ok(f64::from_bits(int_value))
    }

    pub fn read_utf(&mut self) -> Result<String> {
        let length = self.read_short()?;
        let mut buffer = vec![0u8; length as usize];
        self.reader
            .read_exact(&mut buffer)
            .map_err(|_| ConversionError::ReadError("UTF string".to_string()))?;
        String::from_utf8(buffer)
            .map_err(|_| ConversionError::ReadError("UTF string (invalid UTF-8)".to_string()))
    }

    pub fn read_interned_utf(&mut self) -> Result<String> {
        let index = self.read_short()?;
        if index == INTERNED_STRING_NEW_MARKER {
            let string = self.read_utf()?;
            self.interned_strings.push(string.clone());
            Ok(string)
        } else {
            self.interned_strings
                .get(index as usize)
                .cloned()
                .ok_or(ConversionError::InvalidInternedStringIndex(index))
        }
    }

    pub fn read_bytes(&mut self, length: u16) -> Result<Vec<u8>> {
        let mut data = vec![0u8; length as usize];
        self.reader
            .read_exact(&mut data)
            .map_err(|_| ConversionError::ReadError("bytes".to_string()))?;
        Ok(data)
    }
}

// ============================================================================
// Binary XML Deserializer
// ============================================================================

pub struct BinaryXmlDeserializer<R: Read, W: Write> {
    input: DataInput<R>,
    output: W,
}

impl<R: Read, W: Write> BinaryXmlDeserializer<R, W> {
    pub fn new(mut reader: R, output: W) -> Result<Self> {
        let mut magic = [0u8; 4];
        reader
            .read_exact(&mut magic)
            .map_err(|_| ConversionError::ReadError("magic header".to_string()))?;

        if magic != PROTOCOL_MAGIC_VERSION_0 {
            return Err(ConversionError::InvalidMagicHeader {
                expected: PROTOCOL_MAGIC_VERSION_0,
                actual: magic,
            });
        }

        Ok(Self {
            input: DataInput::new(reader),
            output,
        })
    }

    pub fn deserialize(&mut self) -> Result<()> {
        self.output
            .write_all(b"<?xml version=\"1.0\" encoding=\"UTF-8\"?>")?;

        loop {
            match self.process_token() {
                Ok(should_continue) => {
                    if !should_continue {
                        break;
                    }
                }
                Err(ConversionError::ReadError(_)) => {
                    break;
                }
                Err(e) => {
                    eprintln!("Warning: Error parsing token: {}", e);
                    break;
                }
            }
        }

        Ok(())
    }
    fn process_token(&mut self) -> Result<bool> {
        let token = self.input.read_byte()?;
        let command = token & 0x0F;
        let type_info = token & 0xF0;

        match command {
            START_DOCUMENT => Ok(true),
            END_DOCUMENT => Ok(false),
            START_TAG => {
                let tag_name = self.input.read_interned_utf()?;
                self.output.write_all(b"<")?;
                self.output.write_all(tag_name.as_bytes())?;

                while let Ok(next_token) = self.input.peek_byte() {
                    if (next_token & 0x0F) != ATTRIBUTE {
                        break;
                    }

                    let _ = self.input.read_byte()?;
                    self.process_attribute(next_token)?;
                }

                self.output.write_all(b">")?;
                Ok(true)
            }
            END_TAG => {
                let tag_name = self.input.read_interned_utf()?;
                self.output.write_all(b"</")?;
                self.output.write_all(tag_name.as_bytes())?;
                self.output.write_all(b">")?;
                Ok(true)
            }
            TEXT => {
                if type_info == TYPE_STRING {
                    let text = self.input.read_utf()?;
                    if !text.is_empty() {
                        let encoded = encode_xml_entities(&text);
                        self.output.write_all(encoded.as_bytes())?;
                    }
                }
                Ok(true)
            }
            CDSECT => {
                if type_info == TYPE_STRING {
                    let text = self.input.read_utf()?;
                    self.output.write_all(b"<![CDATA[")?;
                    self.output.write_all(text.as_bytes())?;
                    self.output.write_all(b"]]>")?;
                }
                Ok(true)
            }
            COMMENT => {
                if type_info == TYPE_STRING {
                    let text = self.input.read_utf()?;
                    self.output.write_all(b"<!--")?;
                    self.output.write_all(text.as_bytes())?;
                    self.output.write_all(b"-->")?;
                }
                Ok(true)
            }
            PROCESSING_INSTRUCTION => {
                if type_info == TYPE_STRING {
                    let text = self.input.read_utf()?;
                    self.output.write_all(b"<?")?;
                    self.output.write_all(text.as_bytes())?;
                    self.output.write_all(b"?>")?;
                }
                Ok(true)
            }
            DOCDECL => {
                if type_info == TYPE_STRING {
                    let text = self.input.read_utf()?;
                    self.output.write_all(b"<!DOCTYPE ")?;
                    self.output.write_all(text.as_bytes())?;
                    self.output.write_all(b">")?;
                }
                Ok(true)
            }
            ENTITY_REF => {
                if type_info == TYPE_STRING {
                    let text = self.input.read_utf()?;
                    self.output.write_all(b"&")?;
                    self.output.write_all(text.as_bytes())?;
                    self.output.write_all(b";")?;
                }
                Ok(true)
            }
            IGNORABLE_WHITESPACE => {
                if type_info == TYPE_STRING {
                    let text = self.input.read_utf()?;
                    self.output.write_all(text.as_bytes())?;
                }
                Ok(true)
            }
            _ => {
                eprintln!("Warning: Unknown token: {}", command);
                Ok(true)
            }
        }
    }

    fn process_attribute(&mut self, token: u8) -> Result<()> {
        let type_info = token & 0xF0;
        let name = self.input.read_interned_utf()?;

        self.output.write_all(b" ")?;
        self.output.write_all(name.as_bytes())?;
        self.output.write_all(b"=\"")?;

        match type_info {
            TYPE_STRING => {
                let value = self.input.read_utf()?;
                let encoded = encode_xml_entities(&value);
                self.output.write_all(encoded.as_bytes())?;
            }
            TYPE_STRING_INTERNED => {
                let value = self.input.read_interned_utf()?;
                let encoded = encode_xml_entities(&value);
                self.output.write_all(encoded.as_bytes())?;
            }
            TYPE_INT => {
                let value = self.input.read_int()?;
                write!(self.output, "{}", value)?;
            }
            TYPE_INT_HEX => {
                let value = self.input.read_int()?;
                if value == -1 {
                    write!(self.output, "{}", value)?;
                } else {
                    write!(self.output, "{:x}", value as u32)?;
                }
            }
            TYPE_LONG => {
                let value = self.input.read_long()?;
                write!(self.output, "{}", value)?;
            }
            TYPE_LONG_HEX => {
                let value = self.input.read_long()?;
                if value == -1 {
                    write!(self.output, "{}", value)?;
                } else {
                    write!(self.output, "{:x}", value as u64)?;
                }
            }
            TYPE_FLOAT => {
                let value = self.input.read_float()?;
                if value.fract() == 0.0 && value.is_finite() {
                    write!(self.output, "{:.1}", value)?;
                } else {
                    write!(self.output, "{}", value)?;
                }
            }
            TYPE_DOUBLE => {
                let value = self.input.read_double()?;
                if value.fract() == 0.0 && value.is_finite() {
                    write!(self.output, "{:.1}", value)?;
                } else {
                    write!(self.output, "{}", value)?;
                }
            }
            TYPE_BOOLEAN_TRUE => {
                self.output.write_all(b"true")?;
            }
            TYPE_BOOLEAN_FALSE => {
                self.output.write_all(b"false")?;
            }
            TYPE_BYTES_HEX => {
                let length = self.input.read_short()?;
                let bytes = self.input.read_bytes(length)?;
                let hex = hex_string(&bytes);
                self.output.write_all(hex.as_bytes())?;
            }
            TYPE_BYTES_BASE64 => {
                let length = self.input.read_short()?;
                let bytes = self.input.read_bytes(length)?;
                let encoded = base64::engine::general_purpose::STANDARD.encode(&bytes);
                self.output.write_all(encoded.as_bytes())?;
            }
            _ => {
                return Err(ConversionError::UnknownAttributeType(type_info));
            }
        }

        self.output.write_all(b"\"")?;
        Ok(())
    }
}

// ============================================================================
// Converter API
// ============================================================================

pub struct AbxToXmlConverter;

impl AbxToXmlConverter {
    pub fn convert<R: Read, W: Write>(reader: R, writer: W) -> Result<()> {
        let mut deserializer = BinaryXmlDeserializer::new(reader, writer)?;
        deserializer.deserialize()
    }

    pub fn convert_file(input_path: &str, output_path: &str) -> Result<()> {
        if input_path == output_path {
            return Self::convert_file_in_place(input_path);
        }

        let input_file = File::open(input_path)?;
        let reader = BufReader::new(input_file);
        let output_file = File::create(output_path)?;
        let writer = BufWriter::new(output_file);
        Self::convert(reader, writer)
    }

    pub fn convert_stdin_stdout() -> Result<()> {
        let stdin = io::stdin();
        let reader = stdin.lock();
        let stdout = io::stdout();
        let writer = BufWriter::new(stdout.lock());
        Self::convert(reader, writer)
    }

    pub fn convert_stdin_to_file(output_path: &str) -> Result<()> {
        let stdin = io::stdin();
        let reader = stdin.lock();
        let output_file = File::create(output_path)?;
        let writer = BufWriter::new(output_file);
        Self::convert(reader, writer)
    }

    pub fn convert_file_to_stdout(input_path: &str) -> Result<()> {
        let input_file = File::open(input_path)?;
        let reader = BufReader::new(input_file);
        let writer = io::stdout();
        Self::convert(reader, writer)
    }

    fn convert_file_in_place(file_path: &str) -> Result<()> {
        let input_file = File::open(file_path)?;
        let mut reader = BufReader::new(input_file);
        let mut file_data = Vec::new();
        reader.read_to_end(&mut file_data)?;

        let cursor = Cursor::new(file_data);
        let mut output_data = Vec::new();
        {
            let writer = Cursor::new(&mut output_data);
            Self::convert(cursor, writer)?;
        }

        let output_file = File::create(file_path)?;
        let mut writer = BufWriter::new(output_file);
        writer.write_all(&output_data)?;
        writer.flush()?;
        Ok(())
    }

    pub fn convert_bytes(abx_data: &[u8]) -> Result<String> {
        let cursor = Cursor::new(abx_data);
        let mut output_data = Vec::new();
        {
            let writer = Cursor::new(&mut output_data);
            Self::convert(cursor, writer)?;
        }
        String::from_utf8(output_data)
            .map_err(|_| ConversionError::ParseError("Invalid UTF-8 in output".to_string()))
    }

    pub fn convert_vec(abx_data: Vec<u8>) -> Result<String> {
        Self::convert_bytes(&abx_data)
    }
}

// ============================================================================
// CLI
// ============================================================================

struct Cli;

impl Cli {
    fn print_help(program_name: &str) {
        eprintln!("Usage: {} [OPTIONS] <input> [output]", program_name);
        eprintln!();
        eprintln!("Converts Android Binary XML (ABX) to human-readable XML.");
        eprintln!();
        eprintln!("Arguments:");
        eprintln!("  input              Input file path (use '-' for stdin)");
        eprintln!("  output             Output file path (use '-' for stdout)");
        eprintln!("                     If not specified, defaults to stdout or in-place");
        eprintln!();
        eprintln!("Options:");
        eprintln!("  -i, --in-place     Overwrite input file with converted output");
        eprintln!("  -h, --help         Show this help message");
    }

    fn run() -> Result<()> {
        let mut args = env::args();
        let bin_name = args
            .next()
            .as_ref()
            .and_then(|p| std::path::Path::new(p).file_name())
            .and_then(|n| n.to_str())
            .unwrap_or("abx2xml")
            .to_string();

        let args: Vec<String> = args.collect();

        if args.is_empty() || args.iter().any(|a| a == "-h" || a == "--help") {
            Self::print_help(&bin_name);
            std::process::exit(if args.is_empty() { 1 } else { 0 });
        }

        let mut in_place = false;
        let mut input_path = None;
        let mut output_path = None;
        let mut after_double_dash = false;

        for arg in &args {
            if !after_double_dash && arg == "--" {
                after_double_dash = true;
            } else if !after_double_dash && (arg == "-i" || arg == "--in-place") {
                in_place = true;
            } else if input_path.is_none() {
                input_path = Some(arg.as_str());
            } else if output_path.is_none() {
                output_path = Some(arg.as_str());
            } else {
                return Err(ConversionError::ParseError(format!(
                    "Unexpected argument: {}",
                    arg
                )));
            }
        }

        let input_path = input_path.ok_or_else(|| {
            ConversionError::ParseError("Missing required argument: INPUT".to_string())
        })?;

        if in_place && input_path == "-" {
            return Err(ConversionError::ParseError(
                "Cannot use -i option with stdin input".to_string(),
            ));
        }

        let output_path = match output_path {
            Some(path) => path,
            None => {
                if in_place {
                    input_path
                } else {
                    "-"
                }
            }
        };

        match (input_path, output_path) {
            ("-", "-") => AbxToXmlConverter::convert_stdin_stdout(),
            ("-", output) => AbxToXmlConverter::convert_stdin_to_file(output),
            (input, "-") => AbxToXmlConverter::convert_file_to_stdout(input),
            (input, output) => AbxToXmlConverter::convert_file(input, output),
        }
    }
}

fn main() {
    if let Err(e) = Cli::run() {
        eprintln!("Error: {}", e);
        std::process::exit(1);
    }
}
